exonum-java-binding/core/rust/src/testkit/mod.rs

```diff
@@ -213,3 +214,92 @@ fn create_java_keypair<'a>(
         &[secret_key_byte_array.into(), public_key_byte_array.into()],
     )
 }
+
+// Converts Java array of `TestKitServiceInstances` to vector of `InstanceConfig`.
+fn instance_configs_from_java_array(
+    env: &JNIEnv,
+    services: jobjectArray,
+) -> JniResult<Vec<InstanceConfig>> {
+    let mut instance_configs = vec![];
+    let num_services = env.get_array_length(services)?;
+    for i in 0..num_services {
+        let service_obj = env.get_object_array_element(services, i)?;
+
+        let artifact_id = get_string_value(env, service_obj, "artifactId")?;
+        let artifact_filename = get_string_value(env, service_obj, "artifactFilename")?;
+        let service_specs_obj: jobjectArray = env
+            .get_field(
+                service_obj,
+                "serviceSpecs",
+                "[Lcom/exonum/binding/testkit/ServiceSpec;",
+            )?
+            .l()?
+            .into_inner();
+
+        let num_specs = env.get_array_length(service_specs_obj)?;
+
+        for k in 0..num_specs {
+            let service_spec = env.get_object_array_element(service_specs_obj, k)?;
+
+            let (service_id, service_name) = get_service_id_and_name(env, service_spec)?;
+            let config_params: jbyteArray = env
+                .get_field(service_spec, "configuration", "[B")?
+                .l()?
+                .into_inner();
+            let config = env.convert_byte_array(config_params)?;
+
+            let spec = InstanceSpec::new(service_id, service_name, &artifact_id)
+                .expect("Unable to create instance specification for service");
+            let cfg = InstanceConfig::new(spec, Some(artifact_filename.to_bytes()), config);
+            instance_configs.push(cfg);
+        }
+    }
+    Ok(instance_configs)
+}
+
+// Converts Java array of `TimeServiceSpec` to vector of `InstanceCollection`.
+fn time_service_specs_from_java_array(
+    env: &JNIEnv,
+    executor: Executor,
+    service_specs: jobjectArray,
+) -> JniResult<Vec<InstanceCollection>> {
+    let mut instance_collection = vec![];
+    let num_configs = env.get_array_length(service_specs)?;
+    for i in 0..num_configs {
+        let service_spec_obj = env.get_object_array_element(service_specs, i)?;
+
+        let (service_id, service_name) = get_service_id_and_name(env, service_spec_obj)?;
+        let time_provider = env
+            .get_field(
+                service_spec_obj,
+                "timeProvider",
+                "Lcom/exonum/binding/testkit/TimeProviderAdapter;",
+            )?
+            .l()?;
+
+        let provider = JavaTimeProvider::new(executor.clone(), time_provider);
+        let factory =
+            TimeServiceFactory::with_provider(Arc::new(provider) as Arc<dyn TimeProvider>);
+        instance_collection.push(InstanceCollection::new(factory).with_instance(
```

Comment:
In fact, only the latest added instance of SerficeFactory gets mapped to a particular artifact id in case of multiple InstanceCollection per the same artifact.

Reply:
Why doesn't testkit error then? So this code is incorrect?

Reply:
In case of multiple time provider instances - yes. Only the latest instance of the `TimeProviderAdapter` from the `TimeServiceSpec[]` array will be used for all the service instances.

Reply:
Re-written in a way of allowing only one instance of time service in TestKit.